internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,341 @@
+				}))
+			}
+		}
+		endpoints = append(endpoints, resolver.Endpoint{Addresses: addrs})
+	}
+	return addresses, endpoints, true
+}
+
+// updateProxyResolverState updates the proxy resolver state by storing proxy
+// addresses and endpoints, marking the resolver as ready, and triggering a
+// state update if both proxy and target resolvers are ready. If the ClientConn
+// returns a non-nil error, it calls `ResolveNow()` on the target resolver.  It
+// is a StateListener function of wrappingClientConn passed to the proxy resolver.
+func (r *delegatingResolver) updateProxyResolverState(state resolver.State) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if logger.V(2) {
+		logger.Infof("Addresses received from proxy resolver: %s", state.Addresses)
+	}
+	if state.Endpoints != nil {
+		for _, endpoint := range state.Endpoints {
+			r.proxyAddrs = append(r.proxyAddrs, endpoint.Addresses...)
+		}
+	} else {
+		r.proxyAddrs = state.Addresses
+	}
+	r.proxyResolverReady = true
+	addrs, endpoints, ready := r.combinedAddressesIfReadyLocked()
+	// Another possible approach was to block until updates are received from
+	// both resolvers. But this is not used because calling `New()` triggers
+	// `Build()`  for the first resolver, which calls `UpdateState()`. And the
+	// second resolver hasn't sent an update yet, so it would cause `New()` to
+	// block indefinitely.
+	if ready {
+		r.curState.Addresses = addrs
+		r.curState.Endpoints = endpoints
+		if err := r.cc.UpdateState(r.curState); err != nil {
+			r.targetResolver.ResolveNow(resolver.ResolveNowOptions{})
+			return err
+		}
+	}
+	return nil
+}
+
+// updateTargetResolverState updates the target resolver state by storing target
+// addresses, endpoints, and service config, marking the resolver as ready, and
+// triggering a state update if both resolvers are ready. If the ClientConn
+// returns a non-nil error, it calls `ResolveNow()` on the proxy resolver. It
+// is a StateListener function of wrappingClientConn passed to the target resolver.
+func (r *delegatingResolver) updateTargetResolverState(state resolver.State) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if logger.V(2) {
+		logger.Infof("Addresses received from target resolver: %v", state.Addresses)
+	}
+	r.targetResolverState = state
+	r.targetResolverReady = true
+	// Update curState to include other state information, such as the
+	// service config, provided by the target resolver. This ensures
+	// curState contains all necessary information when passed to
+	// UpdateState. The state update is only sent after both the target and
+	// proxy resolvers have sent their updates, and curState has been
+	// updated with the combined addresses.
+	r.curState = state
+	addrs, endpoints, ready := r.combinedAddressesIfReadyLocked()
+	if ready {
+		r.curState.Addresses = addrs
+		r.curState.Endpoints = endpoints
+		if err := r.cc.UpdateState(r.curState); err != nil {
+			r.proxyResolver.ResolveNow(resolver.ResolveNowOptions{})
```

Comment:
I think we want to call `ResolveNow` on the target resolver here.

Reply:
The plan and discussion we had was, if there is an error from the client after sending the update to it, we pass the error directly to resolver from which update state was called and call resolveNow on the other resolver. Like if the update state we receive error on was called from the target resolver, we return the error directly to the target resolver, and call a resolveNow on proxy resolver.

Reply:
I see.